provider/pkg/await/internal/awaiter.go

```diff
@@ -0,0 +1,146 @@
+// Copyright 2024, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package internal
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"github.com/pulumi/pulumi-kubernetes/provider/v4/pkg/await/condition"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+// Awaiter orchestrates a condition Satisfier and optional Observers.
+type Awaiter struct {
+	namespace string
+	condition condition.Satisfier
+	observers []condition.Observer
+}
+
+// NewAwaiter creates a new Awaiter with the given options.
+func NewAwaiter(options ...awaiterOption) (*Awaiter, error) {
+	ea := &Awaiter{}
+	for _, opt := range options {
+		opt.apply(ea)
+	}
+	return ea, nil
+}
+
+// Await blocks until the Condition is met or until the context is canceled.
+// The operation's timeout should be applied to the provided Context.
+func (aw *Awaiter) Await(ctx context.Context) (err error) {
+	if aw.condition == nil {
+		return fmt.Errorf("missing condition")
+	}
+
+	// Start all of our observers. They'll continue until they're canceled.
+	for _, o := range aw.observers {
+		go func(o condition.Observer) {
+			o.Range(func(e watch.Event) bool {
+				_ = o.Observe(e)
+				return true
+			})
+		}(o)
+	}
+
+	// Block until our condition is satisfied, or until our Context is canceled.
+	aw.condition.Range(func(e watch.Event) bool {
+		err = aw.condition.Observe(e)
+		if err != nil {
+			return false
+		}
+		if done, _ := aw.condition.Satisfied(); done {
+			return false
+		}
+		return true
+	})
+
+	// Re-evaluate our condition since its state might have changed during the
+	// iterator's teardown.
+	done, err := aw.condition.Satisfied()
```

Comment:
Is `err` possibly being redefined here?  Am worried that the defer func is writing to the wrong `err`.